File: Challenge/ProblemD.cpp
#include <iostream>
#include <map>
#include <queue>

using namespace std;

class Node {
    public:
        int data;
        Node *left;
        Node *right;
        Node(int d) {
            data = d;
            left = NULL;
            right = NULL;
        }
};

    void topView(Node * root) {
        if (root != nullptr) {
            int distance = 0;
            map<int, int> topNodes;
            map<Node*, int> nodesDis;
            nodesDis[root] = distance;

            queue<Node*> nodes;
            nodes.push(root);

            while (nodes.size()) {
                distance = nodesDis[root];

                if (topNodes.count(distance) == 0) {
                    topNodes[distance] = root->data;
                }
            
                if (root->right) {
                    nodes.push(root->right);
                    nodesDis[root->right] = distance + 1;
                }

                if (root->left) {
                    nodes.push(root->left);
                    nodesDis[root->left] = distance - 1;
                }

                nodes.pop();
                root = nodes.front();
            }

            for (auto i = topNodes.begin(); i != topNodes.end(); i++) {
                cout << i->second << " "; 
            }
        }
        return;
    }

    int main() {
        return 0;
    }

File: Challenge/ProblemD.py
class Node:
    def __init__(self,data):
        self.value = data
        self.distance = 0
        self.left = None
        self.right = None  

def topView(root):
    if (root != None):
        distance = 0
        root.distance = distance
        topNodes = dict()

        nodes = []
        nodes.append(root)
    
        while(len(nodes)):
            root = nodes[0]
            distance = root.distance
    
            if distance not in topNodes:
                topNodes[distance] = root.value
            
            if(root.right):
                nodes.append(root.right)
                root.right.distance = distance + 1

            if(root.left):
                nodes.append(root.left)
                root.left.distance = distance - 1
    
            nodes.pop(0)

        for i in sorted(topNodes):
            print(str(topNodes[i]), end = " ")   
    return  

def main():
    root = Node(1)
    root.left = Node(2)
    root.right = Node(3)
    root.left.right = Node(4)
    root.left.right.right = Node(5)
    root.left.right.right.right = Node(6)
    topView(root)

main()





    